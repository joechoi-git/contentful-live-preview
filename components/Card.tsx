import React from "react";
import Link from "next/link";
import Image from "next/image";
import type { BlogDetail, RelatedBlog } from "@/lib/contentful/api"; // Adjust the import path as necessary
import { isoToFriendlyDate, isoToFriendlyDateTime } from "@/lib/contentful/options"; // Adjust the import path as necessary

interface CardProps {
    blog: RelatedBlog | BlogDetail;
}

const Card: React.FC<CardProps> = ({ blog }) => {
    return blog && blog.slug ? (
        <article className="h-full flex flex-col rounded-lg shadow-lg overflow-hidden">
            <Link href={`/blogs/${blog.slug}`}>
                <Image
                    alt="placeholder"
                    className="aspect-[4/3] object-cover w-full"
                    height="263"
                    src={blog?.heroImage?.url ?? "/default.jpeg"}
                    width="350"
                />
            </Link>
            <div className="flex-1 p-6">
                <Link href={`/blogs/${blog.slug}`}>
                    <h3 className="text-2xl font-bold leading-tight text-zinc-900 dark:text-zinc-50 py-4">
                        {blog.title}
                    </h3>
                </Link>
                <div className="inline-block rounded-full bg-zinc-100 px-3 py-1 text-sm font-semibold text-zinc-800">
                    {blog.categoryName}
                </div>
                <p className="max-w-none text-zinc-500 mt-4 mb-2 text-sm dark:text-zinc-400">
                    {blog.summary}
                </p>
                <p className="max-w-none text-zinc-600 mt-2 mb-2 text-sm font-bold dark:text-zinc-400">
                    Authored By: {blog.author}
                </p>
                <p className="max-w-none text-zinc-600 mt-2 mb-2 text-sm font-bold dark:text-zinc-400">
                    Authored On: {blog.date ? isoToFriendlyDate(blog.date.toString()) : ""}
                </p>
                <p className="max-w-none text-red-600 mt-2 mb-2 text-sm font-bold dark:text-red-400">
                    Published On:{" "}
                    {blog.sys.publishedAt
                        ? isoToFriendlyDateTime(blog.sys.publishedAt.toString())
                        : "In Draft"}
                </p>
                <div className="flex justify-end">
                    <Link
                        className="inline-flex h-10 items-center justify-center text-sm font-medium"
                        href={`/blogs/${blog.slug}`}
                    >
                        Read More â†’
                    </Link>
                </div>
            </div>
        </article>
    ) : null;
};

export default Card;
