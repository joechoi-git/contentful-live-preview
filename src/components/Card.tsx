import React from "react";
import Link from "next/link";
import Image from "next/image";
import type { BlogDetailProps } from "../lib/contentful/adjustedTypes";
import { isoToFriendlyDate, isoToFriendlyDateTime } from "../lib/contentful/Utils";

interface CardProps {
    blog: BlogDetailProps;
}

const Card: React.FC<CardProps> = ({ blog }) => {
    return blog && blog.slug ? (
        <article className="h-full flex flex-col border-2 border-secondary rounded-xl shadow-xl overflow-hidden">
            <Link href={`/blogs/${blog.slug}`}>
                <Image
                    alt="placeholder"
                    className={`aspect-[4/3] object-cover w-full ${blog.sys.publishedAt ? "" : "grayscale"}`}
                    height="263"
                    src={blog?.heroImage?.url ?? "/default.jpeg"}
                    width="350"
                />
            </Link>
            <div className={`flex-1 p-6 ${blog.sys.publishedAt ? "bg-base-100" : "bg-base-300"}`}>
                <Link href={`/blogs/${blog.slug}`}>
                    <h3 className="text-2xl font-bold leading-tight py-4 text-accent">
                        {blog.title}
                    </h3>
                </Link>
                <div className="font-bold">{blog.categoryName}</div>
                <p className="mt-4 mb-2">{blog.summary}</p>
                <p className="mt-2 mb-2 font-bold">Authored By: {blog.author}</p>
                <p className="mt-2 mb-2 font-bold">
                    Authored On: {blog.date ? isoToFriendlyDate(blog.date.toString()) : ""}
                </p>
                <p className="mt-2 mb-2 font-bold text-secondary">
                    Published On:{" "}
                    {blog.sys.publishedAt
                        ? isoToFriendlyDateTime(blog.sys.publishedAt.toString())
                        : "In Draft"}
                </p>
                <div className="flex justify-end">
                    <Link
                        className="inline-flex h-10 items-center justify-center font-medium text-accent"
                        href={`/blogs/${blog.slug}`}
                    >
                        Read More â†’
                    </Link>
                </div>
            </div>
        </article>
    ) : null;
};

export default Card;
