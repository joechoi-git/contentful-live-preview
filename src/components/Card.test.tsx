import React from "react";
import { render, screen } from "@testing-library/react";
import Card from "../components/Card";
import type { RelatedBlog, BlogDetail } from "../lib/contentful/api";

jest.mock("next/link", () => {
    return ({ children }: { children: React.ReactNode }) => children;
});

jest.mock("next/image", () => {
    // eslint-disable-next-line react/display-name, @next/next/no-img-element
    return ({ alt, src }: { alt: string; src: string }) => <img alt={alt} src={src} />;
});

jest.mock("../lib/contentful/options", () => ({
    isoToFriendlyDate: jest.fn((date: string) => date),
    isoToFriendlyDateTime: jest.fn((datetime: string) => datetime)
}));

const mockBlog: RelatedBlog | BlogDetail = {
    slug: "test-slug",
    heroImage: { url: "/test-image.jpg" },
    sys: { id: "1", publishedAt: "2022-01-01T00:00:00Z", publishedVersion: 1 },
    title: "Test Title",
    categoryName: "Test Category",
    summary: "Test Summary",
    author: "Test Author",
    date: "2022-01-01"
};

describe("Card Component", () => {
    test("renders correctly with valid blog data", () => {
        render(<Card blog={mockBlog} />);

        expect(screen.getByRole("img")).toHaveAttribute("src", "/test-image.jpg");
        expect(screen.getByRole("img")).toHaveAttribute("alt", "placeholder");
        expect(screen.getByText("Test Title")).toBeInTheDocument();
        expect(screen.getByText("Test Category")).toBeInTheDocument();
        expect(screen.getByText("Test Summary")).toBeInTheDocument();
        expect(screen.getByText("Authored By: Test Author")).toBeInTheDocument();
        expect(screen.getByText("Authored On: 2022-01-01")).toBeInTheDocument();
        expect(screen.getByText("Published On: 2022-01-01T00:00:00Z")).toBeInTheDocument();
        expect(screen.getByText("Read More â†’")).toBeInTheDocument();
    });

    test("renders null when blog data is not provided", () => {
        const { container } = render(<Card blog={null as unknown as RelatedBlog | BlogDetail} />);
        expect(container).toBeEmptyDOMElement();
    });

    test("matches snapshot", () => {
        const { asFragment } = render(<Card blog={mockBlog} />);
        expect(asFragment()).toMatchSnapshot();
    });
});
