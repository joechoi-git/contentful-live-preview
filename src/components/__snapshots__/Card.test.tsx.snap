// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Card Component renders correctly with given blog props 1`] = `
<DocumentFragment>
  <article
    class="h-full flex flex-col border-2 border-secondary rounded-xl shadow-xl overflow-hidden"
  >
    <a
      href="/blogs/test-blog"
    >
      <img
        alt="placeholder"
        src="https://example.com/image.jpg"
      />
    </a>
    <div
      class="flex-1 p-6 bg-base-100"
    >
      <a
        href="/blogs/test-blog"
      >
        <h3
          class="text-2xl font-bold leading-tight py-4 text-accent"
        >
          Test Blog Title
        </h3>
      </a>
      <div
        class="font-bold"
      >
        Test Category
      </div>
      <p
        class="mt-4 mb-2"
      >
        Test Blog Summary
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        Authored By: John Doe
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        Authored On: January 1, 2023
      </p>
      <p
        class="mt-2 mb-2 font-bold text-secondary"
      >
        Published On: January 1, 2023, 12:00:00 AM UTC
      </p>
      <div
        class="flex justify-end"
      >
        <a
          class="inline-flex h-10 items-center justify-center font-medium text-accent"
          href="/blogs/test-blog"
        >
          Read More →
        </a>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`Card Component renders default image if heroImage.url is not provided 1`] = `
<DocumentFragment>
  <article
    class="h-full flex flex-col border-2 border-secondary rounded-xl shadow-xl overflow-hidden"
  >
    <a
      href="/blogs/test-blog"
    >
      <img
        alt="placeholder"
        src="/default.jpeg"
      />
    </a>
    <div
      class="flex-1 p-6 bg-base-100"
    >
      <a
        href="/blogs/test-blog"
      >
        <h3
          class="text-2xl font-bold leading-tight py-4 text-accent"
        >
          Test Blog Title
        </h3>
      </a>
      <div
        class="font-bold"
      >
        Test Category
      </div>
      <p
        class="mt-4 mb-2"
      >
        Test Blog Summary
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        Authored By: John Doe
      </p>
      <p
        class="mt-2 mb-2 font-bold"
      >
        Authored On: January 1, 2023
      </p>
      <p
        class="mt-2 mb-2 font-bold text-secondary"
      >
        Published On: January 1, 2023, 12:00:00 AM UTC
      </p>
      <div
        class="flex justify-end"
      >
        <a
          class="inline-flex h-10 items-center justify-center font-medium text-accent"
          href="/blogs/test-blog"
        >
          Read More →
        </a>
      </div>
    </div>
  </article>
</DocumentFragment>
`;

exports[`Card Component renders null if blog slug is not provided 1`] = `<DocumentFragment />`;
